src/evm.rs

```diff
@@ -132,6 +168,88 @@ impl<S: Send + Sync + 'static + ForkSource + std::fmt::Debug> EvmNamespaceT
             }
         })
     }
+
+    fn snapshot(&self) -> BoxFuture<Result<U64>> {
+        let inner = Arc::clone(&self.node);
+        let snapshots = Arc::clone(&self.snapshots);
+
+        Box::pin(async move {
+            // validate max snapshots
+            snapshots
+                .read()
+                .map_err(|err| {
+                    log::error!("failed acquiring read lock for snapshot: {:?}", err);
+                    into_jsrpc_error(Web3Error::InternalError)
+                })
+                .and_then(|snapshots| {
+                    if snapshots.len() >= MAX_SNAPSHOTS as usize {
+                        log::error!("maximum number of '{}' snapshots exceeded", MAX_SNAPSHOTS);
+                        Err(into_jsrpc_error(Web3Error::InternalError))
+                    } else {
+                        Ok(())
+                    }
+                })?;
+
+            // snapshot the node
+            let snapshot = inner
+                .read()
+                .map_err(|err| {
+                    format!("failed acquiring read lock to node for snapshot: {:?}", err)
+                })
+                .and_then(|reader| reader.snapshot())
+                .map_err(|err| {
+                    log::error!("failed creating snapshot: {:?}", err);
+                    into_jsrpc_error(Web3Error::InternalError)
+                })?;
+            snapshots
+                .write()
+                .map(|mut snapshots| {
+                    snapshots.push(snapshot);
```

Comment:
suggestion: If we add a log ling notifying the developer that the node has reverted to a snapshot ID, then I think we should also add a log statement when the snapshot is taken
```suggestion
                    log::info!("Reverting node to snapshot {}", snapshots.len() - 1);
                    snapshots.push(snapshot);
```